src/Compilers/Core/Portable/Serialization/StreamObjectReader.cs

```diff
@@ -0,0 +1,1029 @@
+            }
+
+            if (type == typeof(byte))
+            {
+                return Enum.ToObject(enumType, _reader.ReadByte());
+            }
+
+            if (type == typeof(long))
+            {
+                return Enum.ToObject(enumType, _reader.ReadInt64());
+            }
+
+            if (type == typeof(sbyte))
+            {
+                return Enum.ToObject(enumType, _reader.ReadSByte());
+            }
+
+            if (type == typeof(ushort))
+            {
+                return Enum.ToObject(enumType, _reader.ReadUInt16());
+            }
+
+            if (type == typeof(uint))
+            {
+                return Enum.ToObject(enumType, _reader.ReadUInt32());
+            }
+
+            if (type == typeof(ulong))
+            {
+                return Enum.ToObject(enumType, _reader.ReadUInt64());
+            }
+
+            throw ExceptionUtilities.UnexpectedValue(enumType);
+        }
+
+        private Variant ReadObject(EncodingKind kind)
+        {
+            switch (kind)
+            {
+                case EncodingKind.ObjectRef:
+                    return Variant.FromObject(_referenceMap.GetValue(_reader.ReadInt32()));
+                case EncodingKind.ObjectRef_B:
+                    return Variant.FromObject(_referenceMap.GetValue(_reader.ReadByte()));
+                case EncodingKind.ObjectRef_S:
+                    return Variant.FromObject(_referenceMap.GetValue(_reader.ReadUInt16()));
+
+                case EncodingKind.Object_W:
+                    int id = _referenceMap.GetNextReferenceId();
+
+                    Type type = this.ReadType();
+                    uint memberCount = this.ReadCompressedUInt();
+
+                    if (_binder == null)
+                    {
+                        throw NoBinderException(type.FullName);
+                    }
+
+                    var reader = _binder.GetReader(type);
+                    if (reader == null)
+                    {
+                        throw NoReaderException(type.FullName);
+                    }
+
+                    if (memberCount == 0)
+                    {
+                        return ConstructObject(type, (int)memberCount, reader, id);
+                    }
+                    else
+                    {
+                        _consumerStack.Push(new Consumer(type, (int)memberCount, _valueStack.Count, reader, id));
+                        return Variant.None;
+                    }
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(kind);
+            }
+        }
+
+        private Variant ConstructObject(Type type, int memberCount, Func<ObjectReader, object> reader, int id)
+        {
+            _memberReader.Reset();
+
+            // take members from the stack
+            for (int i = 0; i < memberCount; i++)
+            {
+                _memberReader.List.Add(_valueStack.Pop());
```

Comment:
I want the values to be processed (read/written) from the stream in the same order they were written by the object instance.